Allow lookups based on short identifiers

iverberk opened the pull request:
This change introduces the ability to specify identifiers for jobs, allocs, evals and nodes on the command line with as little as one or two characters, provided that it uniquely identifies the resource. An error with the possible results will be provided when the short identifier has multiple results.

This is the first part of an implementation for https://github.com/hashicorp/nomad/issues/54. Truncating the UUID will be implemented separatly. This pull-request requires changes in go-memdb and go-immutable-radix that are specified in pull-request https://github.com/hashicorp/go-immutable-radix/pull/3 and https://github.com/hashicorp/go-memdb/pull/8.

dadgar commented:
Hey @iverberk, I closed your other PRs for immutable-radix and go-memdb. They both have the constructs we need to accomplish this already. Going to leave some comments to guide how we can do this.

dadgar commented on nomad/state/state_store.go:
```diff
@@ -241,13 +241,26 @@ func (s *StateStore) UpdateNodeDrain(index uint64, nodeID string, drain bool) er
 
 	if existing != nil {
-		return existing.(*structs.Node), nil
+		// Return exact match directly
```
So it is best if we revert all these to their original state and then we can add additional methods such as `NodeByIDPrefix`. These methods can return a memdb.ResultIterator and use memdb.Txn().Get() and use the "_prefix" operator.

We can then add HTTP endpoints that would take a prefix and would call these new methods. And then they could return back a fixed set to the client. This would then let the user refine the search if multiple are returned or if just one is returned, the cli could just call with the full ID.

dadgar commented:
Let me know if you have any questions

iverberk commented:
Hi Alex, thanks for the feedback. Great to hear that the prefix searching is already supported by the db libraries, I overlooked it somehow. So before I update the pull-request, a couple of questions:

1. From quickly looking at the memdb code I gather that the indexer must implement the PrefixFromArgs method in order to be searchable by prefix. Currently this is only implemented by the StringFieldIndex and CompoundIndex. This means that jobs and nodes will be prefix-searchable, but the allocs and evals won't be because they use the UUIDFieldIndex, which does not implement the PrefixIndexer. Is this correct and if so, how would you suggest we go about this? I was also wondering if there is a specific reason why the indexer type for the node identifier is not also an UUIDFieldIndex?

2. It's not entirely clear to me what you want the end-user interaction to look like. The original goal was to have short identifiers for the cli 'node-status', 'eval-monitor' and 'alloc-status' commands. So the user would enter the first few characters of the identifier and it would execute the command if it uniquely identified the resource. Otherwise an error would be returned with possible options. Is this indeed the desired interaction? If so, in which method would the additional HTTP endpoints be called? 

I'll start working on the implementation tomorrow. If you could answer these questions for me that would be great, thanks!

dadgar commented:
1) Yes we will have to make UUIDFieldIndexer meet that interface. I will add that for you. The reason is node id is user specifiable.

2) So with alloc-status and eval-monitor you can just detect if the string is less than 36 characters (since it is a UUID) and then call the new endpoint to get a list of the possible matches. If that returns a single match you can call the GetEval or GetAlloc method with the exact UUID, otherwise display the options to the user.

For node-status, it is a little tricker. You could assume exact and if there is an error try the prefix endpoint, or if you wanted to be fancy you could do the requests in parallel.

dadgar commented:
@iverberk: I made the UUIDFieldIndex support PrefixFromArgs. So you are good to go on that front.

iverberk commented:
The branch has been updated according to the proposed design. It also incorporates a change to display short node identifiers when possible. Please let me know what you think and if anything needs changing.

iverberk commented:
The tests seem to fail but I'm unable to even make the tests pass on master, either on OSX or in the included Vagrant virtual machine. So I think the errors are actually not related to the changes in this pull-request but I might be wrong. Please let me know if tests break due to changes in this PR.

dadgar commented on command/node_status.go:
```diff
@@ -160,3 +164,27 @@ func (c *NodeStatusCommand) Run(args []string) int {
 	}
 	return 0
 }
+
```
I would remove both these methods and replace them with one that takes the original nodes and returns a map of their original ids to their shortened ones.

`shortIds(nodes []*api.NodeListStub) map[string]string`

With this method I think you could be a bit more advanced with how you shorted. You could start at 8 characters and if there is a collision, then you up it to 9, etc, until you get no collisions. Further it is probably still worth while to shorten the non-custom ones.

dadgar commented on command/node_status.go:
```diff
@@ -78,12 +80,14 @@ func (c *NodeStatusCommand) Run(args []string) int {
 			return 0
 		}
 
+		shortenNodeId := shouldShortenNodeIds(nodes)
```
There should probably be a verbose flag that does not shorten the node ids

dadgar commented on nomad/alloc_endpoint.go:
```diff
@@ -80,9 +81,40 @@ func (a *Alloc) GetAlloc(args *structs.AllocSpecificRequest,
+					return err
+				}
+			} else {
+				iter, err := snap.AllocByIDPrefix(args.AllocID)
```
So as I mentioned earlier we don't want to put this logic here (for all the endpoints). These endpoint are for looking up the particular id.

I think the best way to add this feature is to add a `?prefix=true` parameter to the list requests (there is one for job/eval/alloc).

You would have to add:
* parsing [here](https://github.com/hashicorp/nomad/blob/master/command/agent/http.go#L271) and [example call](https://github.com/hashicorp/nomad/blob/master/command/agent/job_endpoint.go#L23)
* add to struct [here](https://github.com/hashicorp/nomad/blob/master/nomad/structs/structs.go#L58)
* update list methods to use the new state store calls you added. [Example](https://github.com/hashicorp/nomad/blob/master/nomad/job_endpoint.go#L260)
   * So if prefix is set use `JobByIdPrefix` instead of `Jobs`
* The CLI commands would then need to first call with the specific ID and if there is an error call list with the prefix and then show the output to the user.

This has the benefit that it makes the list endpoint more usable, maintains the API of the get endpoints, makes everything much more testable, and we aren't using the error as a return value.

iverberk commented:
Hi Alex, thanks for your patience in guiding me through this. I've updated the pull-request based on your comments. I've only updated the node status-command fully because I want to verify the changes before I implement the other commands similarly. I've removed the truncation from this pull-request, it will be done in another pull-request to not mix things up to much. Tests seem to be happy as well now. Please let me know if the current changes are in line with your thoughts!

iverberk commented on command/node_drain.go:
```diff
@@ -68,8 +68,15 @@ func (c *NodeDrainCommand) Run(args []string) int {
+	// Check if node exists
+	node, _, err := client.Nodes().Info(nodeID, nil)
+	if err != nil {
+		c.Ui.Error(fmt.Sprintf("Error toggling drain mode: %s", err))
```
This is strictly not an error for toggling the drain mode, but rather for looking up the node. However, the testcase checks for this string so I'd like to hear your thoughts on this.

dadgar replied:
I think this is fine as the `err` will be something like "node with id "foobar" not found"

iverberk commented on nomad/structs/structs.go:
```diff
@@ -69,6 +69,9 @@ type QueryOptions struct {
 	AllowStale bool
+
+	// If set, used as prefix for resource list searches
+	Prefix string
```
I created it as a string and not as a boolean, as that would require an additional parameter with the actual prefix string.

dadgar replied:
Looks good

iverberk commented on api/api.go:
```diff
@@ -31,6 +31,9 @@ type QueryOptions struct {
 	WaitTime time.Duration
+
+	// If set, used as prefix for resource list searches
+	Prefix string
```
It is also added in the structs. Once this PR is done I will take a look at removing the api specific structs and using those in the structs package as specified in issue #329 .

iverberk commented on command/node_status.go:
```diff
@@ -100,8 +102,41 @@ func (c *NodeStatusCommand) Run(args []string) int {
+			return 0
+		}
+		//  Query full node information for unique prefix match
+		node, _, err = client.Nodes().Info(nodes[0].ID, nil)
```
The extra lookup is required since the list function does not return the full information that we need, kinda annoying to make three api calls but I can't see another way?

dadgar replied:
Yeah since we return the stub you have to. Its okay tho, its the CLI, it isn't high QPS where performance matters too much

dadgar commented:
@iverberk: Great work! This is exactly what I was envisioning! Hope you also agree that is cleaner!

dadgar commented on command/node_status.go:
```diff
@@ -100,8 +102,41 @@ func (c *NodeStatusCommand) Run(args []string) int {
+		}
+		// Return error if no nodes are found
+		if len(nodes) == 0 {
+			c.Ui.Error(fmt.Sprintf("Node not found"))
```
I would make this more verbose. `fmt.Sprintf("No node(s) with prefix %q found", nodeID)`

dadgar commented:
Also down the line please update the docs. For example [here](https://github.com/hashicorp/nomad/blob/master/website/source/docs/http/nodes.html.md)

iverberk commented:
Another big update. This mostly just adds similar functionality to the allocations, evaluations and jobs. I had to add and modify tests to cover all the new branches. I will add another commit to update the docs. Let me know what you think and merry christmas! :-)

iverberk commented:
@dadgar: no idea why that particular test is failing. I can't reproduce it locally (all tests are passing, including the erroneous one). Can you help me investigate?

iverberk commented:
I've added updates to the documentation. Feel free to adjust or comment on wording. The documentation updates seemed to have fixed the tests as well....

dadgar commented:
@iverberk: Awesome work! I will try to look at it soon! Just got back from vacation so lots to get through. The reason the tests weren't passing for a bit was from an upstream bug that is fixed now!

dadgar commented on api/nodes_test.go:
```diff
@@ -38,6 +38,52 @@ func TestNodes_List(t *testing.T) {
+	})
+
+	// Find node based on four character prefix
+	testutil.WaitForResult(func() (bool, error) {
```
Why does this have to be wrapped in a WaitForResult?

iverberk replied:
That's an oversight on my part, fixed.

dadgar commented on command/agent/alloc_endpoint_test.go:
```diff
@@ -53,6 +53,52 @@ func TestHTTP_AllocsList(t *testing.T) {
+			t.Fatalf("missing last contact")
+		}
+
+		// Check the job
```
Check the alloc

dadgar commented on command/agent/alloc_endpoint_test.go:
```diff
@@ -53,6 +53,52 @@ func TestHTTP_AllocsList(t *testing.T) {
+
+		// Check the job
+		n := obj.([]*structs.AllocListStub)
+		if len(n) != 1 {
```
I would also test that you get the right ID

dadgar commented on command/agent/eval_endpoint_test.go:
```diff
@@ -53,6 +53,52 @@ func TestHTTP_EvalList(t *testing.T) {
+		}
+
+		// Check the job
+		e := obj.([]*structs.Evaluation)
```
Same comments here

dadgar commented on command/agent/node_endpoint_test.go:
```diff
@@ -56,6 +56,55 @@ func TestHTTP_NodesList(t *testing.T) {
+			t.Fatalf("missing last contact")
+		}
+
+		// Check the job
```
nodes

dadgar commented on command/alloc_status.go:
```diff
@@ -68,8 +68,36 @@ func (c *AllocStatusCommand) Run(args []string) int {
-		return 1
+		allocs, _, err := client.Allocations().PrefixList(allocID)
+		if err != nil {
+			c.Ui.Error(fmt.Sprintf("Error querying allocation: %s", err))
```
Can you use %v when formatting errors.

dadgar commented on command/alloc_status.go:
```diff
@@ -68,8 +68,36 @@ func (c *AllocStatusCommand) Run(args []string) int {
+			c.Ui.Output(formatList(out))
+			return 0
+		}
+		alloc, _, err = client.Allocations().Info(allocs[0].ID, nil)
```
I would put a comment saying:

`// Prefix lookup matched a single allocation`

dadgar commented on command/node_drain.go:
```diff
@@ -68,8 +68,47 @@ func (c *NodeDrainCommand) Run(args []string) int {
+		if len(nodes) > 1 {
+			// Format the nodes list that matches the prefix so that the user
+			// can create a more specific request
+			out := make([]string, len(nodes)+1)
```
It may be worth adding an output before saying something along the lines of:
"Please disambiguate the desired node: "

Something like this probably makes sense for eval-monitor too, or really anything were the expected behavior isn't executed because of ambiguity.

iverberk replied:
I'm not sure how you feel about the way I put it in the code with two newlines. I didn't see a nicer way to display the help (perhaps by expanding the array and adding the line(s) in there). I really think there should be a line between the disambiguation message and the list.

dadgar replied:
Yeah I think that is fine. And I agree that it should have a blank in-between.

dadgar commented on website/source/docs/commands/alloc-status.html.md.erb:
```diff
@@ -19,8 +19,8 @@ current state of its tasks.
 
-An allocation ID must be provided. This specific allocation will be queried
-and detailed information for it will be dumped.
+An allocation ID (prefix) must be provided. This specific allocation will be
```
I would update these to say something along the lines of:

`An allocation ID or prefix must be provided. If there is an exact match, the full details of the allocation will be displayed. Otherwise, a list of matching allocations and information will be displayed.`

dadgar commented:
So excited for this @iverberk! Thank you so much! I left some simple comments but it looks ready to merge :+1:

iverberk commented:
It think I addressed your comments with the new commit. Please take a look and let me know! :-)

dadgar commented:
Awesome! Thank you so much @iverberk, this is really great!

iverberk commented:
@dadgar really happy with this merged PR! Thanks so much for reviewing and guidance.
